sway-core/src/semantic_analysis/syntax_tree.rs

```diff
@@ -262,6 +267,82 @@ impl TypedParseTree {
+/// satisfied. We're basically checking the following condition:
+///    if trait B is implemented for type T, then trait A_i is also implemented for type T for
+///    every A_i such that A_i is a supertrait of B.
+///
+/// This nicely works for transitive supertraits as well.
+///
+fn check_supertraits(
+    typed_tree_nodes: &[TypedAstNode],
+    namespace: &NamespaceRef,
+) -> Vec<CompileError> {
+    let mut errors = vec![];
+    for node in typed_tree_nodes {
+        if let TypedAstNodeContent::Declaration(TypedDeclaration::ImplTrait {
+            trait_name,
+            span,
+            type_implementing_for,
+            ..
+        }) = &node.content
+        {
+            if let CompileResult {
+                value: Some(TypedDeclaration::TraitDeclaration(tr)),
+                ..
+            } = namespace.get_call_path(trait_name)
+            {
+                let supertraits = tr.supertraits;
+                for supertrait in &supertraits {
+                    let impl_of_supertrait = typed_tree_nodes.iter().find(|search_node| {
+                        if let TypedAstNodeContent::Declaration(TypedDeclaration::ImplTrait {
+                            trait_name: search_node_trait_name,
+                            type_implementing_for: search_node_type_implementing_for,
+                            ..
+                        }) = &search_node.content
+                        {
+                            if let (
+                                CompileResult {
+                                    value: Some(TypedDeclaration::TraitDeclaration(tr1)),
+                                    ..
+                                },
+                                CompileResult {
+                                    value: Some(TypedDeclaration::TraitDeclaration(tr2)),
+                                    ..
+                                },
+                            ) = (
+                                namespace.get_call_path(search_node_trait_name),
+                                namespace.get_call_path(&supertrait.name),
+                            ) {
+                                return (tr1.name == tr2.name)
+                                    && (type_implementing_for
+                                        == search_node_type_implementing_for);
+                            }
+                        }
+                        false
+                    });
+                    if impl_of_supertrait.is_none() {
+                        // The two errors below should really be a single error (and a "note"),
+                        // but we don't have a way today to point to two separate locations in the
+                        // user code with a single error.
+                        errors.push(CompileError::SupertraitImplMissing {
+                            supertrait_name: supertrait.name.span().span.as_str().to_string(),
```

Comment:
This is a bit crazy!  Is this the same?
```suggestion
                            supertrait_name: supertrait.name.to_string(),
```
`CallPath` implements `Display` and I think includes the prefix + suffix.

Reply:
Indeed! Your suggestions is a lot better.